daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -185,96 +224,84 @@ class VariableMap private(vTable: Map[GlobalQName, List[List[Variable]]])
-        firstTier match {
-
-          case Variable(VariableDefined, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
-
-          case Variable(VariableUndefined, DataValue.NoValue, ctxt, defaultExpr) :: rest => {
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
-
-          case Variable(VariableSet, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            referringContext.SDE("Cannot set variable %s twice. State was: %s. Existing value: %s", ctxt.globalQName, VariableSet, v)
-          }
-
-          case Variable(VariableRead, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            // referringContext.SDE
-            pstate.SDW("Cannot set variable %s after reading the default value. State was: %s. Existing value: %s", ctxt.globalQName, VariableSet, v)
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
+        case VariableRead => {
+          referringContext.SDE("Cannot set variable %s after reading the default value. State was: %s. Existing value: %s",
+          variable.rd.globalQName, VariableSet, variable.value)
+        }
 
-          case _ => Assert.invariantFailed("variable map internal list structure not as expected: " + x)
+        case _ => {
+          variable.setValue(VariableUtils.convert(newValue.getAnyRef.toString, variable.rd))
+          variable.setState(VariableSet)
         }
       }
-      case x => Assert.invariantFailed("variables data structure not as expected. Should not be " + x)
     }
   }
 
+  /**
+   * Creates a new instance of a variable
+   */
+  def newVariableInstance(vrd: VariableRuntimeData) = {
+    val varQName = vrd.globalQName
+    val stack = vTable.get(varQName)
+    if (stack.isDefined) {
+      stack.get.push(vrd.newVariableInstance)
+    }
+  }
+
+  def removeVariableInstance(vrd: VariableRuntimeData) {
+    val varQName = vrd.globalQName
+    val stack = vTable.get(varQName)
+    if (stack.isDefined) {
+      stack.get.pop
+    }
+  }
+
+
   private lazy val externalVarGlobalQNames: Seq[GlobalQName] =
     vTable.map { pair => pair match {
-      case (gqn, firstTier :: _) => firstTier match {
-        case (variable@Variable(_, v, ctxt, _)) :: _ if (ctxt.external) => variable.rd.globalQName
+      case (gqn, stack) if (stack.top.rd.external) => {
+        stack.top.rd.globalQName
       }
-      case (_, Nil) => Assert.invariantFailed("empty tier")
+      case (_, stack) if (stack.isEmpty) => Assert.invariantFailed("empty tier")
```

Comment:
Does stack.top throw an exception if the stack is empty? Seems like it's never possible to hit this case.